python/ray/data/_internal/logical/rules/operator_fusion.py

```diff
@@ -460,6 +477,77 @@ def fused_all_to_all_transform_fn(
         # Return the fused physical operator.
         return op
 
+    @classmethod
+    def _can_fuse_map_ops(
+        cls,
+        upstream_op: AbstractMap,
+        downstream_op: AbstractMap,
+        context: "DataContext",
+    ) -> bool:
+        ds_bundle_min_rows_req = downstream_op._min_rows_per_bundled_input
+        us_bundle_min_rows_req = upstream_op._min_rows_per_bundled_input
+
+        # Do not fuse read op with downstream map op in case when downstream has
+        # `min_rows_per_input_bundle` specified (to avoid reducing reading parallelism)
+        if upstream_op.is_read_op() and ds_bundle_min_rows_req is not None:
```

Comment:
In addition to read ops, I think we should disable fusion as long as the previous map op doesn't preserve num rows (e.g., read, filter, map_batches, flat_map, etc)

Reply:
for map_batches, typically it preserves num rows. 
But today we don't enforce that. 
related issue https://github.com/ray-project/ray/issues/36295
One option is to enforce that by default, and add a flag to allow violation.

Reply:
> One option is to enforce that by default, and add a flag to allow violation.

I don't think we can do that anymore with our public API -- i can totally see that being too limiting.

Regardless, though preserving num-rows for proper limit push-downs is an important topic but tangential to this change.